pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java

```diff
@@ -181,56 +183,85 @@ public CompletableFuture<List<CompletableFuture<SchemaAndMetadata>>> getAllSchem
+                                        checkCompatibilityFuture.complete(null);
+                                    }
+                                    return checkCompatibilityFuture.thenCompose(v -> {
+                                        byte[] context = hashFunction.hashBytes(schema.getData()).asBytes();
+                                        SchemaRegistryFormat.SchemaInfo info = SchemaRegistryFormat.SchemaInfo
+                                                .newBuilder()
+                                                .setType(Functions.convertFromDomainType(schema.getType()))
+                                                .setSchema(ByteString.copyFrom(schema.getData()))
+                                                .setSchemaId(schemaId)
+                                                .setUser(schema.getUser())
+                                                .setDeleted(false)
+                                                .setTimestamp(clock.millis())
+                                                .addAllProps(toPairs(schema.getProps()))
+                                                .build();
+
+                                        long start = this.clock.millis();
+                                        schemaStorage
+                                                .put(schemaId, info.toByteArray(), context)
+                                                .thenAccept(sv -> {
+                                                    if (log.isDebugEnabled()) {
+                                                        log.debug("[{}] Schema is successfully added", schemaId);
+                                                    }
+                                                    this.stats.recordPutLatency(schemaId, this.clock.millis() - start);
+                                                    persistentFuture.complete(sv);
+                                                })
+                                                .exceptionally(t -> {
+                                                    if (t != null) {
+                                                        if (t.getCause() instanceof AlreadyExistsException
+                                                                || t.getCause() instanceof BadVersionException) {
+                                                            // retry if put schemaLocator to zk failed caused by
+                                                            // race condition
+                                                            if (log.isDebugEnabled()) {
+                                                                log.debug("[{}] Put schema failed because of {}, retry",
+                                                                        t.getCause().getMessage(), schemaId);
+                                                            }
+                                                            putSchemaIfAbsent(schemaId, schema, strategy)
```

Comment:
Call `putSchemaIfAbsent` here will retry once, but it will not record metrics `this.stats.recordPutLatency` or `this.stats.recordPutFailed`. Maybe we can add a new method `putSchemaWithRetryIfAbsent`, this new method only wraps the original method `putSchemaIfAbsent`

Reply:
+1

Reply:
> Call putSchemaIfAbsent here will retry once, but it will not record metrics this.stats.recordPutLatency or this.stats.recordPutFailed. 

`this.stats.recordPutLatenc`  is recorded incorrectly since `long start = this.clock.millis()` ;
`this.stats.recordPutFailed` has been recorded when exceptions (not `AlreadyExistsException` or `BadVersionException`) happened, so no need to record again.

> Maybe we can add a new method putSchemaWithRetryIfAbsent, this new method only wraps the original method putSchemaIfAbsent

Agree, and add a new metric to indicate `putOpsRetryCounter` retry times